src/coinsviewcacheasync.h

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_COINSVIEWCACHEASYNC_H
+#define BITCOIN_COINSVIEWCACHEASYNC_H
+
+#include <attributes.h>
+#include <coins.h>
+#include <primitives/block.h>
+#include <primitives/transaction.h>
+#include <tinyformat.h>
+#include <uint256.h>
+#include <util/threadnames.h>
+
+#include <algorithm>
+#include <atomic>
+#include <barrier>
+#include <cstdint>
+#include <optional>
+#include <ranges>
+#include <thread>
+#include <utility>
+#include <vector>
+
+static constexpr uint32_t WORKER_THREADS{4};
+
+/**
+ * CCoinsViewCache subclass that asynchronously fetches block inputs in parallel.
+ * Only used in ConnectBlock to pass as an ephemeral view that can be reset if the block is invalid.
+ * It provides the same interface as CCoinsViewCache, overriding the FetchCoin private method and Flush.
+ * It adds an additional StartFetching method to provide the block, and Reset to reset the state if Flush is not called.
+ *
+ * The cache spawns a fixed set of worker threads that fetch Coins for each input in a block.
+ * When FetchCoin() is called, the main thread waits for the corresponding coin to be fetched and returns it.
+ * While waiting, the main thread will also fetch coins to maximize parallelism.
+ *
+ * Worker threads are synchronized with the main thread using a barrier, which is used at the beginning of fetching to
+ * start the workers and at the end to ensure all workers have finished before the next block is started.
+ */
+class CoinsViewCacheAsync : public CCoinsViewCache
+{
+private:
+    //! The latest input not yet being fetched. Workers atomically increment this when fetching.
+    mutable std::atomic_uint32_t m_input_head{0};
+    //! The latest input not yet accessed by a consumer. Only the main thread increments this.
+    mutable uint32_t m_input_tail{0};
+
+    //! The inputs of the block which is being fetched.
+    struct InputToFetch {
+        //! Workers set this after setting the coin. The main thread tests this before reading the coin.
+        std::atomic_flag ready{};
+        //! The outpoint of the input to fetch;
+        const COutPoint& outpoint;
+        //! The coin that workers will fetch and main thread will insert into cache.
+        std::optional<Coin> coin{std::nullopt};
+
+        /**
+         * We only move when m_inputs reallocates during setup.
+         * We never move after work begins, so we don't have to copy other members.
+         */
+        InputToFetch(InputToFetch&& other) noexcept : outpoint{other.outpoint} {}
+        explicit InputToFetch(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
```

Comment:
[Sonar is complaining](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&branch=31132-69310ec0035351e486cfb51cb66492639f6f8b47&id=aureleoules_bitcoin&open=AZrV9y4tqc704xdUcnxE&tab=more_info) that we're violating the Rule of Five here, it's a bit verbose, but maybe we could:
```suggestion
        InputToFetch(InputToFetch&& other) noexcept : outpoint{other.outpoint} {}
        explicit InputToFetch(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
        InputToFetch(const InputToFetch&) = delete;
        InputToFetch& operator=(const InputToFetch&) = delete;
        InputToFetch& operator=(InputToFetch&&) = delete;
        ~InputToFetch() = default;
```